File: src/Functional/OCaml/builder.ml
type report_builder = {
  reset : unit -> unit;
  add_title : string -> unit;
  add_section : string -> string -> unit;
  build : unit -> string;
}

let make_builder html =
  let parts = ref [] in
  let reset () = parts := [] in
  let add value = parts := !parts @ [value] in
  let add_title title = if html then add ("<h1>" ^ title ^ "</h1>") else add ("# " ^ title) in
  let add_section heading body =
    if html then (add ("<h2>" ^ heading ^ "</h2>"); add ("<p>" ^ body ^ "</p>"))
    else (add ("## " ^ heading); add body)
  in
  let build () = String.concat (if html then "" else "\n") !parts in
  { reset; add_title; add_section; build }

let build_availability_report builder =
  builder.reset ();
  builder.add_title "Service status";
  builder.add_section "Availability" "99.95%";
  builder.build ()

let () =
  print_endline (build_availability_report (make_builder false));
  print_endline "---";
  print_endline (build_availability_report (make_builder true))

File: src/Functional/Scala/Builder.scala
trait ReportBuilder {
  def reset(): Unit
  def addTitle(title: String): Unit
  def addSection(heading: String, body: String): Unit
  def build(): String
}

final class TextReportBuilder extends ReportBuilder {
  private val parts = scala.collection.mutable.ArrayBuffer.empty[String]
  def reset(): Unit = parts.clear()
  def addTitle(title: String): Unit = parts += s"# $title"
  def addSection(heading: String, body: String): Unit = {
    parts += s"## $heading"
    parts += body
  }
  def build(): String = parts.mkString("\n")
}

final class HtmlReportBuilder extends ReportBuilder {
  private val parts = scala.collection.mutable.ArrayBuffer.empty[String]
  def reset(): Unit = parts.clear()
  def addTitle(title: String): Unit = parts += s"<h1>$title</h1>"
  def addSection(heading: String, body: String): Unit = {
    parts += s"<h2>$heading</h2>"
    parts += s"<p>$body</p>"
  }
  def build(): String = parts.mkString
}

object Builder {
  def buildAvailabilityReport(builder: ReportBuilder): String = {
    builder.reset()
    builder.addTitle("Service status")
    builder.addSection("Availability", "99.95%")
    builder.build()
  }

  def main(args: Array[String]): Unit = {
    println(buildAvailabilityReport(new TextReportBuilder))
    println("---")
    println(buildAvailabilityReport(new HtmlReportBuilder))
  }
}
